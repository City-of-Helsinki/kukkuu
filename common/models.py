import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.managers import TranslatableQuerySet as ParlerTranslatableQuerySet
from parler.models import TranslatableModel as ParlerTranslatableModel


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True


class TranslatableQuerySet(ParlerTranslatableQuerySet):
    def create_translatable_object(self, **kwargs):
        translations = kwargs.pop("translations")
        obj = self.create(**kwargs)
        for translation in translations:
            language_code = translation.pop("language_code")
            obj.create_translation(language_code=language_code, **translation)
        return obj


class TranslatableModel(ParlerTranslatableModel):
    objects = TranslatableQuerySet.as_manager()

    class Meta:
        abstract = True
