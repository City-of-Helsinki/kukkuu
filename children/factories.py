import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
