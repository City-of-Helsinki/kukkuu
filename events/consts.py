class NotificationType:
    EVENT_PUBLISHED = "event_published"
    EVENT_GROUP_PUBLISHED = "event_group_published"
    OCCURRENCE_ENROLMENT = "occurrence_enrolment"
    OCCURRENCE_UNENROLMENT = "occurrence_unenrolment"
    OCCURRENCE_CANCELLED = "occurrence_cancelled"
    OCCURRENCE_REMINDER = "occurrence_reminder"
