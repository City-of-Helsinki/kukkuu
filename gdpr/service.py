import logging
from typing import TYPE_CHECKING, Optional

from helsinki_gdpr.types import ErrorResponse

from subscriptions.models import FreeSpotNotificationSubscription

if TYPE_CHECKING:
    from users.models import User as UserType

logger = logging.getLogger(__name__)


def get_user(user: "UserType") -> "UserType":
    """Function used by the Helsinki Profile GDPR API to get the "user"
    instance from the "GDPR Model"
    instance. Since in our case the GDPR Model
    and the user are one and the same, we simply return
    the same User instance that is given as a parameter.

    Args:
        user (UserType): the User instance whose GDPR data is being queried

    Returns:
        UserType: the same User instance

    References:
        https://github.com/City-of-Helsinki/helsinki-profile-gdpr-api/blob/808dcd30a745f6d18cdf36ccaf07b0cd25844ab0/README.md.
        https://github.com/City-of-Helsinki/kerrokantasi/blob/256134d4049f0bd1598d59caaebbb813be2d7d9c/kerrokantasi/gdpr.py.
    """
    logger.info(f"GDPR data request called for user '{user.uuid}'.")
    return user


def clear_data(user: "UserType", dry_run: bool) -> Optional[ErrorResponse]:
    """Function used by the Helsinki Profile GDPR API to clear GDPR data fields
    related to the user. The User instance or
    any related model instances won't be deleted,
    but the user will be disabled and the user data and
    the related data will be anonymised.
    The GDPR API package will run this within a transaction.

    Args:
        user (UserType): the User instance to be deleted along with related GDPR data
        dry_run (bool): a boolean telling if this is a dry run of the function or not

    Returns:
        Optional[ErrorResponse]: any errors that occured,
            e.g. the object might not exist or the data deletion might be denied.

    References:
        https://github.com/City-of-Helsinki/helsinki-profile-gdpr-api/blob/808dcd30a745f6d18cdf36ccaf07b0cd25844ab0/README.md.
        https://github.com/City-of-Helsinki/kerrokantasi/blob/256134d4049f0bd1598d59caaebbb813be2d7d9c/kerrokantasi/gdpr.py.
    """
    from users.models import Guardian

    logger.info(f"GDPR data clear called for user '{user.uuid}'.")
    user.clear_gdpr_sensitive_data_fields()
    try:
        user.guardian.clear_gdpr_sensitive_data_fields()
        for child in user.guardian.children.all():
            child.clear_gdpr_sensitive_data_fields()
            FreeSpotNotificationSubscription.objects.filter(child=child).delete()

    except Guardian.DoesNotExist:
        logger.warning(
            "Could not call 'clear_gdpr_sensitive_data_fields' for a guardian object, "
            f"since there is no guardian linked to the user {user.uuid}."
        )
