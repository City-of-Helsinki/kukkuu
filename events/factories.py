import factory
import pytz

from events.models import Event, Occurrence
from venues.factories import VenueFactory


class EventFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("text", max_nb_chars=64)
    duration = factory.Faker("random_int", max=300)
    short_description = factory.Faker("text", max_nb_chars=64)
    description = factory.Faker("text")
    image = factory.Faker("file_name", extension="jpg")
    participants_per_invite = factory.Faker(
        "random_element", elements=[x[0] for x in Event.PARTICIPANT_AMOUNT_CHOICES]
    )
    capacity_per_occurrence = factory.Faker("random_int", max=1000)

    class Meta:
        model = Event


class OccurrenceFactory(factory.django.DjangoModelFactory):
    time = factory.Faker("date_time", tzinfo=pytz.timezone("Europe/Helsinki"))
    event = factory.SubFactory(EventFactory)
    venue = factory.SubFactory(VenueFactory)

    class Meta:
        model = Occurrence
