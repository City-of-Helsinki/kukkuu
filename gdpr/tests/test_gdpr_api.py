import urllib.parse
import uuid
from typing import Optional

import pytest
import requests_mock
from django.conf import settings
from django.contrib.auth import get_user_model
from django.urls import reverse
from django.utils import timezone
from rest_framework.test import APIClient

from children.factories import ChildWithGuardianFactory
from events.factories import EnrolmentFactory, TicketSystemPasswordFactory
from gdpr.consts import CLEARED_VALUE
from gdpr.service import clear_data
from gdpr.tests.conftest import get_api_token_for_user_with_scopes
from subscriptions.factories import FreeSpotNotificationSubscriptionFactory
from subscriptions.models import FreeSpotNotificationSubscription
from users.factories import GuardianFactory, UserFactory
from users.models import Guardian

User = get_user_model()


def request_gdpr_delete(
    user,
    id_value,
    scopes=(settings.GDPR_API_DELETE_SCOPE,),
    query_params=None,
    data=None,
):
    gdpr_api_client = APIClient()

    with requests_mock.Mocker() as req_mock:
        auth_header = get_api_token_for_user_with_scopes(user, scopes, req_mock)
        gdpr_api_client.credentials(HTTP_AUTHORIZATION=auth_header)

        if query_params:
            query = "?" + urllib.parse.urlencode(query_params)
        else:
            query = ""

        request_kwargs = {"format": "json"}
        if data:
            request_kwargs["data"] = data

        return gdpr_api_client.delete(
            reverse(
                "helsinki_gdpr:gdpr_v1",
                kwargs={settings.GDPR_API_MODEL_LOOKUP: id_value},
            )
            + query,
            **request_kwargs,
        )


def assert_clear(user, original_password: Optional[str] = None):
    if original_password:
        assert user.password != original_password
    assert user.first_name == ""
    assert user.last_name == ""
    assert user.username
    try:
        guardian = user.guardian
        child = guardian.children.first()
        assert guardian.first_name == CLEARED_VALUE
        assert guardian.last_name == CLEARED_VALUE
        assert child.name == ""
    except Guardian.DoesNotExist:
        pass


def delete_user(user, params):
    assert User.objects.count() == 1
    response = request_gdpr_delete(user, user.uuid, **params)
    assert response.status_code == 204
    assert User.objects.count() == 1


@pytest.mark.django_db
def test_clear_data_service():
    child = ChildWithGuardianFactory(name="Karate Kid")
    FreeSpotNotificationSubscriptionFactory(child=child)
    guardian = child.guardians.all().first()
    user = guardian.user
    original_password = user.password
    clear_data(user=user, dry_run=False)
    user.refresh_from_db()
    assert_clear(user, original_password)
    assert FreeSpotNotificationSubscription.objects.count() == 0


@pytest.mark.django_db
@pytest.mark.parametrize("key", ["data", "query_params"])
def test_gdpr_api_delete_profile_dry_run(true_value, user, key):
    email = user.email
    assert email != ""
    delete_user(user, {key: {"dry_run": true_value}})
    user.refresh_from_db()
    assert user.email == email
    assert User.objects.count() == 1


@pytest.mark.django_db
@pytest.mark.parametrize("key", ["data", "query_params"])
def test_gdpr_api_delete_profile(false_value, user, key):
    original_password = user.password
    delete_user(user, {key: {"dry_run": false_value}})
    user.refresh_from_db()
    assert_clear(user, original_password)


@pytest.mark.django_db
def test_gdpr_api_delete_profile_no_params(user):
    original_password = user.password
    delete_user(user, {})
    user.refresh_from_db()
    assert_clear(user, original_password)


@pytest.mark.django_db
def test_gdpr_api_user_not_found(user):
    """The response with a status code 204 is given in the case
    that the service does not contain any data for the profile
    or is completely unaware of the identified profile.

    See more: https://profile-api.dev.hel.ninja/docs/gdpr-api/
    """
    assert User.objects.count() == 1
    response = request_gdpr_delete(user=user, id_value=uuid.uuid4())
    assert response.status_code == 204
    assert User.objects.count() == 1


@pytest.mark.django_db
def test_get_profile_data_from_gdpr_api(
    snapshot, gdpr_api_client, requests_mock, project
):
    guardian = GuardianFactory(
        id=uuid.UUID("8dff3da4-a329-4b81-971a-bc509df679b1"),
        user__uuid=uuid.UUID("fa354000-3c0c-11eb-86c5-acde48001122"),
    )
    child = ChildWithGuardianFactory(relationship__guardian=guardian)
    EnrolmentFactory.create_batch(5, child=child)
    TicketSystemPasswordFactory.create_batch(5, child=child)
    guardian = child.guardians.first()

    user = guardian.user
    user.administered_projects = [project]
    user.last_login = timezone.now()
    user.save()
    auth_header = get_api_token_for_user_with_scopes(
        user, [settings.GDPR_API_QUERY_SCOPE], requests_mock
    )
    gdpr_api_client.credentials(HTTP_AUTHORIZATION=auth_header)
    response = gdpr_api_client.get(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: user.uuid},
        )
    )
    assert response.status_code == 200
    snapshot.assert_match(response.json())


@pytest.mark.django_db
def test_delete_profile_data_from_gdpr_api(user, gdpr_api_client, requests_mock):
    auth_header = get_api_token_for_user_with_scopes(
        user, [settings.GDPR_API_DELETE_SCOPE], requests_mock
    )
    gdpr_api_client.credentials(HTTP_AUTHORIZATION=auth_header)
    response = gdpr_api_client.delete(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: user.uuid},
        )
    )
    assert response.status_code == 204
    with pytest.raises(User.DoesNotExist):
        User.objects.get(username=user.username)


@pytest.mark.django_db
def test_gdpr_api_requires_authentication(user, gdpr_api_client):
    response = gdpr_api_client.get(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: user.uuid},
        )
    )
    assert response.status_code == 401

    response = gdpr_api_client.delete(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: user.uuid},
        )
    )
    assert response.status_code == 401


@pytest.mark.django_db
def test_user_can_only_access_his_own_profile(user, gdpr_api_client, requests_mock):
    auth_header = get_api_token_for_user_with_scopes(
        user,
        [settings.GDPR_API_QUERY_SCOPE, settings.GDPR_API_DELETE_SCOPE],
        requests_mock,
    )
    gdpr_api_client.credentials(HTTP_AUTHORIZATION=auth_header)

    another_user = UserFactory()
    response = gdpr_api_client.get(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: another_user.uuid},
        )
    )
    assert response.status_code == 403

    response = gdpr_api_client.delete(
        reverse(
            "helsinki_gdpr:gdpr_v1",
            kwargs={settings.GDPR_API_MODEL_LOOKUP: another_user.uuid},
        )
    )
    assert response.status_code == 403
