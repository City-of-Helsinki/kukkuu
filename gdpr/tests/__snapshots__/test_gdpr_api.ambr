# serializer version: 1
# name: test_get_profile_data_from_gdpr_api
  dict({
    'children': list([
      dict({
        'key': 'UUID',
        'value': 'fa354000-3c0c-11eb-86c5-acde48001122',
      }),
      dict({
        'key': 'USERNAME',
        'value': 'jeffersonkimberly',
      }),
      dict({
        'key': 'FIRST_NAME',
        'value': 'Alexis',
      }),
      dict({
        'key': 'LAST_NAME',
        'value': 'Black',
      }),
      dict({
        'key': 'EMAIL',
        'value': 'michellewalker@example.net',
      }),
      dict({
        'key': 'ADMINISTERED_PROJECTS',
        'value': list([
        ]),
      }),
      dict({
        'key': 'LAST_LOGIN',
        'value': '2020-12-12T00:00:00+00:00',
      }),
      dict({
        'key': 'DATE_JOINED',
        'value': '2020-12-12T00:00:00+00:00',
      }),
      dict({
        'children': list([
          dict({
            'key': 'ID',
            'value': '8dff3da4-a329-4b81-971a-bc509df679b1',
          }),
          dict({
            'key': 'USER',
            'value': 'Black Alexis (michellewalker@example.net)',
          }),
          dict({
            'key': 'FIRST_NAME',
            'value': 'Michael',
          }),
          dict({
            'key': 'LAST_NAME',
            'value': 'Patton',
          }),
          dict({
            'key': 'EMAIL',
            'value': 'michellewalker@example.net',
          }),
          dict({
            'key': 'PHONE_NUMBER',
            'value': '235.857.7767x124',
          }),
          dict({
            'key': 'HAS_ACCEPTED_COMMUNICATION',
            'value': False,
          }),
          dict({
            'children': list([
              dict({
                'children': list([
                  dict({
                    'key': 'NAME',
                    'value': 'Katherine Gomez',
                  }),
                  dict({
                    'key': 'BIRTHYEAR',
                    'value': 2023,
                  }),
                  dict({
                    'key': 'POSTAL_CODE',
                    'value': '49763',
                  }),
                  dict({
                    'children': list([
                      dict({
                        'children': list([
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'TIME',
                                'value': '1981-12-29T15:08:25.080157+00:00',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Increase thank certainly again thought summer.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'EVENT_GROUP',
                                    'value': None,
                                  }),
                                  dict({
                                    'children': list([
                                      dict({
                                        'key': 'YEAR',
                                        'value': 2020,
                                      }),
                                      dict({
                                        'key': 'NAME_WITH_TRANSLATIONS',
                                        'value': dict({
                                          'en': '',
                                          'fi': 'Testiprojekti',
                                          'sv': '',
                                        }),
                                      }),
                                    ]),
                                    'key': 'PROJECT',
                                  }),
                                  dict({
                                    'key': 'TICKET_SYSTEM',
                                    'value': 'internal',
                                  }),
                                ]),
                                'key': 'EVENT',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Clearly middle moment strong hand.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'ADDRESS_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': '''
                                        779 Kevin Isle Suite 550
                                        Jonathanburgh, AK 06327
                                      ''',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'VENUE',
                              }),
                            ]),
                            'key': 'OCCURRENCE',
                          }),
                        ]),
                        'key': 'ENROLMENT',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'TIME',
                                'value': '1973-04-20T19:00:51.908992+00:00',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Democratic focus significant kind various laugh.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'EVENT_GROUP',
                                    'value': None,
                                  }),
                                  dict({
                                    'children': list([
                                      dict({
                                        'key': 'YEAR',
                                        'value': 2020,
                                      }),
                                      dict({
                                        'key': 'NAME_WITH_TRANSLATIONS',
                                        'value': dict({
                                          'en': '',
                                          'fi': 'Testiprojekti',
                                          'sv': '',
                                        }),
                                      }),
                                    ]),
                                    'key': 'PROJECT',
                                  }),
                                  dict({
                                    'key': 'TICKET_SYSTEM',
                                    'value': 'internal',
                                  }),
                                ]),
                                'key': 'EVENT',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Medical check word style also. Question national throw three.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'ADDRESS_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': '''
                                        Unit 7738 Box 4455
                                        DPO AE 93387
                                      ''',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'VENUE',
                              }),
                            ]),
                            'key': 'OCCURRENCE',
                          }),
                        ]),
                        'key': 'ENROLMENT',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'TIME',
                                'value': '1983-12-14T10:21:44.139987+00:00',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Policy sport available.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'EVENT_GROUP',
                                    'value': None,
                                  }),
                                  dict({
                                    'children': list([
                                      dict({
                                        'key': 'YEAR',
                                        'value': 2020,
                                      }),
                                      dict({
                                        'key': 'NAME_WITH_TRANSLATIONS',
                                        'value': dict({
                                          'en': '',
                                          'fi': 'Testiprojekti',
                                          'sv': '',
                                        }),
                                      }),
                                    ]),
                                    'key': 'PROJECT',
                                  }),
                                  dict({
                                    'key': 'TICKET_SYSTEM',
                                    'value': 'internal',
                                  }),
                                ]),
                                'key': 'EVENT',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Person material alone throughout investment check increase.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'ADDRESS_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': '''
                                        18631 David Wells Apt. 018
                                        South Mikeburgh, NH 37426
                                      ''',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'VENUE',
                              }),
                            ]),
                            'key': 'OCCURRENCE',
                          }),
                        ]),
                        'key': 'ENROLMENT',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'TIME',
                                'value': '1989-06-15T21:38:14.498478+00:00',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Month score father middle brother station physical very.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'EVENT_GROUP',
                                    'value': None,
                                  }),
                                  dict({
                                    'children': list([
                                      dict({
                                        'key': 'YEAR',
                                        'value': 2020,
                                      }),
                                      dict({
                                        'key': 'NAME_WITH_TRANSLATIONS',
                                        'value': dict({
                                          'en': '',
                                          'fi': 'Testiprojekti',
                                          'sv': '',
                                        }),
                                      }),
                                    ]),
                                    'key': 'PROJECT',
                                  }),
                                  dict({
                                    'key': 'TICKET_SYSTEM',
                                    'value': 'internal',
                                  }),
                                ]),
                                'key': 'EVENT',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Moment apply president unit positive.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'ADDRESS_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': '''
                                        7921 Fischer Ranch
                                        Port Timothy, MI 90086
                                      ''',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'VENUE',
                              }),
                            ]),
                            'key': 'OCCURRENCE',
                          }),
                        ]),
                        'key': 'ENROLMENT',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'TIME',
                                'value': '1980-04-10T05:46:28.089342+00:00',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Suggest claim PM they. South blue reach ask.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'EVENT_GROUP',
                                    'value': None,
                                  }),
                                  dict({
                                    'children': list([
                                      dict({
                                        'key': 'YEAR',
                                        'value': 2020,
                                      }),
                                      dict({
                                        'key': 'NAME_WITH_TRANSLATIONS',
                                        'value': dict({
                                          'en': '',
                                          'fi': 'Testiprojekti',
                                          'sv': '',
                                        }),
                                      }),
                                    ]),
                                    'key': 'PROJECT',
                                  }),
                                  dict({
                                    'key': 'TICKET_SYSTEM',
                                    'value': 'internal',
                                  }),
                                ]),
                                'key': 'EVENT',
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Process who catch bad. Seven process determine call.',
                                      'sv': '',
                                    }),
                                  }),
                                  dict({
                                    'key': 'ADDRESS_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': '''
                                        1235 Eric Route Suite 932
                                        New Manuelshire, PR 28866
                                      ''',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'VENUE',
                              }),
                            ]),
                            'key': 'OCCURRENCE',
                          }),
                        ]),
                        'key': 'ENROLMENT',
                      }),
                    ]),
                    'key': 'ENROLMENTS',
                  }),
                  dict({
                    'children': list([
                      dict({
                        'children': list([
                          dict({
                            'key': 'ASSIGNED_AT',
                            'value': None,
                          }),
                          dict({
                            'key': 'VALUE',
                            'value': 'jR1NWxHl(x',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'NAME_WITH_TRANSLATIONS',
                                'value': dict({
                                  'en': '',
                                  'fi': 'Half state four hear trouble among face three.',
                                  'sv': '',
                                }),
                              }),
                              dict({
                                'key': 'EVENT_GROUP',
                                'value': None,
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'YEAR',
                                    'value': 2020,
                                  }),
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Testiprojekti',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'PROJECT',
                              }),
                              dict({
                                'key': 'TICKET_SYSTEM',
                                'value': 'internal',
                              }),
                            ]),
                            'key': 'EVENT',
                          }),
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                        ]),
                        'key': 'TICKETSYSTEMPASSWORD',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'ASSIGNED_AT',
                            'value': None,
                          }),
                          dict({
                            'key': 'VALUE',
                            'value': ')1q1TpGzpE',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'NAME_WITH_TRANSLATIONS',
                                'value': dict({
                                  'en': '',
                                  'fi': 'Where official agree order just raise.',
                                  'sv': '',
                                }),
                              }),
                              dict({
                                'key': 'EVENT_GROUP',
                                'value': None,
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'YEAR',
                                    'value': 2020,
                                  }),
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Testiprojekti',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'PROJECT',
                              }),
                              dict({
                                'key': 'TICKET_SYSTEM',
                                'value': 'internal',
                              }),
                            ]),
                            'key': 'EVENT',
                          }),
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                        ]),
                        'key': 'TICKETSYSTEMPASSWORD',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'ASSIGNED_AT',
                            'value': None,
                          }),
                          dict({
                            'key': 'VALUE',
                            'value': '+2(bkVx7(8',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'NAME_WITH_TRANSLATIONS',
                                'value': dict({
                                  'en': '',
                                  'fi': 'Many happy better agree concern often almost.',
                                  'sv': '',
                                }),
                              }),
                              dict({
                                'key': 'EVENT_GROUP',
                                'value': None,
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'YEAR',
                                    'value': 2020,
                                  }),
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Testiprojekti',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'PROJECT',
                              }),
                              dict({
                                'key': 'TICKET_SYSTEM',
                                'value': 'internal',
                              }),
                            ]),
                            'key': 'EVENT',
                          }),
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                        ]),
                        'key': 'TICKETSYSTEMPASSWORD',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'ASSIGNED_AT',
                            'value': None,
                          }),
                          dict({
                            'key': 'VALUE',
                            'value': '&KXlatF20E',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'NAME_WITH_TRANSLATIONS',
                                'value': dict({
                                  'en': '',
                                  'fi': 'Program song couple central each color.',
                                  'sv': '',
                                }),
                              }),
                              dict({
                                'key': 'EVENT_GROUP',
                                'value': None,
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'YEAR',
                                    'value': 2020,
                                  }),
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Testiprojekti',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'PROJECT',
                              }),
                              dict({
                                'key': 'TICKET_SYSTEM',
                                'value': 'internal',
                              }),
                            ]),
                            'key': 'EVENT',
                          }),
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                        ]),
                        'key': 'TICKETSYSTEMPASSWORD',
                      }),
                      dict({
                        'children': list([
                          dict({
                            'key': 'ASSIGNED_AT',
                            'value': None,
                          }),
                          dict({
                            'key': 'VALUE',
                            'value': '_O2oUJuK@V',
                          }),
                          dict({
                            'children': list([
                              dict({
                                'key': 'NAME_WITH_TRANSLATIONS',
                                'value': dict({
                                  'en': '',
                                  'fi': 'Bit other she chair cover whether.',
                                  'sv': '',
                                }),
                              }),
                              dict({
                                'key': 'EVENT_GROUP',
                                'value': None,
                              }),
                              dict({
                                'children': list([
                                  dict({
                                    'key': 'YEAR',
                                    'value': 2020,
                                  }),
                                  dict({
                                    'key': 'NAME_WITH_TRANSLATIONS',
                                    'value': dict({
                                      'en': '',
                                      'fi': 'Testiprojekti',
                                      'sv': '',
                                    }),
                                  }),
                                ]),
                                'key': 'PROJECT',
                              }),
                              dict({
                                'key': 'TICKET_SYSTEM',
                                'value': 'internal',
                              }),
                            ]),
                            'key': 'EVENT',
                          }),
                          dict({
                            'key': 'CHILD',
                            'value': 'Katherine Gomez (2023)',
                          }),
                        ]),
                        'key': 'TICKETSYSTEMPASSWORD',
                      }),
                    ]),
                    'key': 'TICKET_SYSTEM_PASSWORDS',
                  }),
                  dict({
                    'children': list([
                    ]),
                    'key': 'FREE_SPOT_NOTIFICATION_SUBSCRIPTIONS',
                  }),
                  dict({
                    'key': 'NOTES',
                    'value': 'GDPR sensitive test notes for child',
                  }),
                ]),
                'key': 'CHILD',
              }),
            ]),
            'key': 'CHILDREN',
          }),
        ]),
        'key': 'GUARDIAN',
      }),
    ]),
    'key': 'USER',
  })
# ---
